docs/sources/developers/plugins/working-with-data-frames.md

```diff
@@ -0,0 +1,109 @@
+
+// Create data frame from values.
+const frame = toDataFrame({
+    name: "http_requests_total",
+    fields: [
+        { name: "Time", type: FieldType.time, values: timeValues },
+        { name: "Value", type: FieldType.number, values: numberValues }
+    ]
+});
+```
+
+> **Note:** Data frames representing time series contain at least a `time` field, and a `number` field. By convention, built-in plugins use `Time` and `Value` as field names for data frames containing time series data.
+
+As you can see from the example, creating data frames like this requires that your data is already stored as columnar data. If you already have the records in the form of an array of objects, you can pass it to `toDataFrame` which tries to guess the schema based on the types and names of the objects in the array. If you're creating complex data frames this way, make sure to verify that you get the schema you expect.
+
+```ts
+const series = [
+    { Time: 1599471973065, Value: 12.3 },
+    { Time: 1599471975729, Value: 28.6 }
+];
+
+const frame = toDataFrame(series);
+frame.name = 'http_requests_total'
+```
+
+## Read values from a data frame
+
+When you're building a panel plugin, the data frames returned by the data source are available from the `data` prop in your panel component.
+
+```ts
+const SimplePanel: React.FC<Props> = ({ data }) => {
+  const frame = data.series[0];
+
+  // ...
+}
+```
+
+Before you start reading the data, think about what data you expect. For time series, we can expect a `Time` and a `Value` field, so it's safe to hard code them.
+
+```ts
+const timeField = frame.fields.find(field => field.name === 'Time');
+if (timeField.type !== FieldType.time) {
+  throw new Error('Invalid type for Time field')
+}
+
+const valueField = frame.fields.find(field => field.name === 'Value');
+if (valueField.type !== FieldType.number) {
+  throw new Error('Invalid type for Value field')
+}
+```
+
+Consider other types of visualizations may need multiple dimensions. For example, a bubble chart that uses three numeric fields: The X-axis, Y-axis, and one for the radius of each bubble. In this case, instead of hard coding the field names, we recommend that you let the user choose the field to use for each dimension.
+
+```ts
+const x = frame.fields.find(field => field.name === options.xField);
+const y = frame.fields.find(field => field.name === options.yField);
+const size = frame.fields.find(field => field.name === options.sizeField);
+
+for (let i = 0; i < frame.length; i++) {
+  const row = [x.get(i), y.get(i), size.get(i)];
+}
+```
+
+Alternatively, you can use the [DataFrameView]({{< relref "../../packages_api/data/dataframeview.md" >}}) which gives you an array of objects that contain a property for each field in the frame.
+
+```ts
+const view = new DataFrameView(frame);
+
+view.forEach(row => {
+  console.log(row[options.xField], row[options.yField], row[options.sizeField]);
+})
+```
+
+## Convert legacy formats to data frames
```

Comment:
This seems like an important task. I suggest either putting it first (since it converts old data into the required format for the example to work!) or putting it in a separate task topic by itself so that it has more visibility. If you choose the latter route, then you might expand the instructions a bit.

Reply:
It's helpful if you're maintaining an older plugin, but not relevant to anyone starting out today I don't think this should be at the top of this document. I could see this living in the migration guide, but I believe this document is where users will look first.

Reply:
I thought lots of people were converting from pre-7.0 to now. It certainly comes up a lot in #plugins. 

So...as a person who might need to update an old plugin, "working with data frames" does not necessarily lead me to "here's how to convert my old plugin data to the new scheme."